apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/KeysTests.cs

```diff
@@ -187,42 +205,126 @@ public void CreateKeySetFromKeys()
-                k => Assert.Equal("k3", k.KeyParts.Values[0].StringValue)
+                k => Assert.Equal("k1", k.ToProtobuf(s_conversionOptions).Values[0].StringValue),
+                k => Assert.Equal("k2", k.ToProtobuf(s_conversionOptions).Values[0].StringValue),
+                k => Assert.Equal("k3", k.ToProtobuf(s_conversionOptions).Values[0].StringValue)
             );
         }
 
+        [Fact]
+        public void TypeMappings()
+        {
+            // Without specifying the type mapping.
+            var builder = new SpannerConnectionStringBuilder();
+            var keySet = KeySet.FromParameters(new SpannerParameterCollection { new SpannerParameter { ParameterName = "k1", Value = 3.14m }, 
+                new SpannerParameter { ParameterName = "k2", Value = new DateTime(2022, 06, 08) } });
+            var defaultProtobufValues = keySet.ToProtobuf(builder.ConversionOptions);
+            Assert.Equal(3.14d, defaultProtobufValues.Keys[0].Values[0].NumberValue); // default to double.
+            Assert.Equal("2022-06-08T00:00:00Z", defaultProtobufValues.Keys[0].Values[1].StringValue); // default to Timestamp.
+
+            // Specifying the type mappings.
+            builder.ClrToSpannerTypeDefaultMappings = "DecimalToNumeric,DateTimeToDate";
+            defaultProtobufValues = keySet.ToProtobuf(builder.ConversionOptions);
+            Assert.Equal("3.14", defaultProtobufValues.Keys[0].Values[0].StringValue); // Numeric string.
+            Assert.Equal("2022-06-08", defaultProtobufValues.Keys[0].Values[1].StringValue); // Date string.
+
+            // Revert the type mappings.
+            builder.ClrToSpannerTypeDefaultMappings = "";
+            defaultProtobufValues = keySet.ToProtobuf(builder.ConversionOptions);
+            Assert.Equal(3.14d, defaultProtobufValues.Keys[0].Values[0].NumberValue); // default to double.
+            Assert.Equal("2022-06-08T00:00:00Z", defaultProtobufValues.Keys[0].Values[1].StringValue); // default to Timestamp.
+        }
+
         [Fact]
         public void CreateKeySetFromRanges()
         {
             var ranges = KeySet.FromRanges(
                 KeyRange.ClosedOpen(new Key("begin"), new Key("end")), KeyRange.ClosedClosed(new Key(1L), new Key(2L)));
             Assert.Collection(ranges.Ranges,
-                range => Assert.Equal(KeyRange.ClosedOpen(new Key("begin"), new Key("end")), range),
-                range => Assert.Equal(KeyRange.ClosedClosed(new Key(1L), new Key(2L)), range)
+                range => Assert.Equal(KeyRange.ClosedOpen(new Key("begin"), new Key("end")).ToProtobuf(s_conversionOptions), range.ToProtobuf(s_conversionOptions)),
+                range => Assert.Equal(KeyRange.ClosedClosed(new Key(1L), new Key(2L)).ToProtobuf(s_conversionOptions), range.ToProtobuf(s_conversionOptions))
             );
         }
 
+        [Fact]
+        public void ToStringKeyWithBuilder()
```

Comment:
Probably after this PR, I think it would be worth having one test class per production class. I don't mind whether they're in the same file or not though.